swesmith/bug_gen/adapters/typescript.py

```diff
@@ -0,0 +1,275 @@
+                    return first_line[:brace_pos].strip()
+                else:
+                    result = first_line.strip()
+                    if result.endswith(";"):
+                        result = result[:-1].strip()
+                    return result
+        return self.src_code.split("\n")[0].strip()
+
+    @property
+    def stub(self) -> str:
+        signature = self.signature
+        if self.node.type == "class_declaration":
+            return f"{signature} {{\n\t// {TODO_REWRITE}\n}}"
+        elif self.node.type == "arrow_function":
+            if "=>" in signature:
+                return f"{signature} {{\n\t// {TODO_REWRITE}\n}}"
+            else:
+                return f"{signature} => {{\n\t// {TODO_REWRITE}\n}}"
+        else:
+            return f"{signature} {{\n\t// {TODO_REWRITE}\n}}"
+
+    @property
+    def complexity(self) -> int:
+        def walk(node):
+            score = 0
+            if node.type in [
+                "if_statement",
+                "else_clause",
+                "for_statement",
+                "for_in_statement",
+                "for_of_statement",
+                "while_statement",
+                "do_statement",
+                "switch_statement",
+                "case_clause",
+                "catch_clause",
+                "conditional_expression",
+            ]:
+                score += 1
+            if node.type == "binary_expression":
+                for child in node.children:
+                    if hasattr(child, "text") and child.text.decode("utf-8") in [
+                        "&&",
+                        "||",
+                    ]:
+                        score += 1
+            for child in node.children:
+                score += walk(child)
+            return score
+
+        return 1 + walk(self.node)
+
+
+def get_entities_from_file_ts(
+    entities: list[TypeScriptEntity],
+    file_path: str,
+    max_entities: int = -1,
+) -> list[TypeScriptEntity]:
+    """
+    Parse a .ts file and return up to max_entities top-level functions and classes.
+    If max_entities < 0, collects them all.
+    """
+    parser = Parser(TS_LANGUAGE)
+    try:
+        file_content = open(file_path, "r", encoding="utf8").read()
+    except UnicodeDecodeError:
+        warnings.warn(f"Could not decode file {file_path}", stacklevel=2)
+        return entities
+    tree = parser.parse(bytes(file_content, "utf8"))
+    root = tree.root_node
+    lines = file_content.splitlines()
+    _walk_and_collect_ts(root, entities, lines, str(file_path), max_entities)
+    return entities
+
+
+def _walk_and_collect_ts(node, entities, lines, file_path, max_entities):
+    if 0 <= max_entities == len(entities):
+        return
+    if node.type == "ERROR":
+        warnings.warn(f"Error encountered parsing {file_path}", stacklevel=2)
+        return
+    # Only collect classes, functions, and methods (not variables)
+    if node.type == "export_statement":
```

Comment:
Is the case handling `export_statement` necessary here? It seems like we will recurse down the tree and capture the entities anyway if it was removed?